src/MQTTClient.php

```diff
@@ -143,44 +155,114 @@ public function connect(
-                STREAM_CLIENT_CONNECT,
-                $socketContext
-            );
-        } else {
-            $connectionString = 'tcp://' . $this->getHost() . ':' . $this->getPort();
+        $contextOptions = [];
+
+        if ($useTls) {
+            $contextOptions['ssl'] = [];
+
+            if ($this->settings->shouldTlsVerifyPeer() || $this->hasCertificateAuthorityFile()) {
+                // As it does not make any sense to specify a certificate file without
+                // enabling peer verification, we automatically enable it in that case
+                $contextOptions['ssl']['verify_peer'] = true;
+            } else {
+                // You get a warning output for free, on every connection
+                $this->logger->warning('Enabled TLS encryption without peer verification - DANGEROUS SECURITY ISSUE');
+                $contextOptions['ssl']['verify_peer'] = false;
+            }
 
-            $this->socket = stream_socket_client(
+            $contextOptions['ssl']['verify_peer_name'] = $this->settings->shouldTlsVerifyPeerName();
+
+            if ($this->hasCertificateAuthorityFile()) {
+                $this->logger->info(sprintf('Using certificate authority file [%s] to verify peer name.', $this->getCertificateAuthorityFile()));
+                $contextOptions['ssl']['cafile'] = $this->getCertificateAuthorityFile();
+            }
+        }
+
+        $connectionString = 'tcp://' . $this->getHost() . ':' . $this->getPort();
+        $socketContext    = stream_context_create($contextOptions);
+        $socket           = @stream_socket_client(
+            $connectionString,
+            $errorCode,
+            $errorMessage,
+            $this->settings->getSocketTimeout(),
+            STREAM_CLIENT_CONNECT,
+            $socketContext
+        );
+
+        if ($socket === false) {
+            $this->logger->error(sprintf(
+                'Establishing a connection with the MQTT broker using connection string [%s] failed: %s (code %d).',
                 $connectionString,
-                $errorCode,
                 $errorMessage,
-                60,
-                STREAM_CLIENT_CONNECT
+                $errorCode
+            ));
+            // errorCode is a POSIX socket error (0-255)
+            throw new ConnectingToBrokerFailedException(
+                self::EXCEPTION_CONNECTION_SOCKET_ERROR,
+                "Socket error [$errorCode]: " . $errorMessage,
+                $errorMessage,
+                (string) $errorCode
             );
         }
 
-        if ($this->socket === false) {
-            $this->logger->error(sprintf('Establishing a connection with the MQTT broker using connection string [%s] failed.', $connectionString));
-            throw new ConnectingToBrokerFailedException($errorCode, $errorMessage);
+        if ($useTls) {
+            error_clear_last();
+            $retval = @stream_socket_enable_crypto($socket, true, STREAM_CRYPTO_METHOD_ANY_CLIENT);
```

Comment:
The PHP documentation has the following to say about the third parameter of this function:
> If omitted, the `crypto_method` context option on the stream's SSL context will be used instead.

Would it make sense to make this configurable (i.e. to avoid using TLSv1_x)? We could simply put it in the context configuration above.